Spring 08

engineer027 opened the pull request.

engineer027 committed:
add role model dao and service

engineer027 committed:
Update modet User, his dao, service, and mapper

engineer027 committed:
create CustomUserDetailsService and update another classes

engineer027 committed:
create InjectData and add new dependency in pom

Sofasmile commented on src/main/java/com/dev/cinema/config/SecurityConfig.java:
```diff
@@ -1,27 +1,32 @@
-                .inMemoryAuthentication()
-                .passwordEncoder(getEncoder())
-                .withUser("bob").password(getEncoder().encode("1234")).roles("USER");
+                .userDetailsService((userDetailsService));
```
we should add here passwordEncoder

Sofasmile commented on src/main/java/com/dev/cinema/config/SecurityConfig.java:
```diff
@@ -1,27 +1,32 @@
     protected void configure(HttpSecurity httpSecurity) throws Exception {
         httpSecurity
                 .authorizeRequests()
+                .antMatchers(HttpMethod.POST, "/register").permitAll()
```
let's add more urks here

Sofasmile commented on src/main/java/com/dev/cinema/controllers/InjectData.java:
```diff
@@ -0,0 +1,25 @@
+import javax.annotation.PostConstruct;
+import org.springframework.stereotype.Controller;
+
+@Controller
+public class InjectData {
```
let's rename this class to DataInitializer and use the annotation Component instead of Controller

Sofasmile commented on src/main/java/com/dev/cinema/controllers/InjectData.java:
```diff
@@ -0,0 +1,25 @@
+        roleService.add(user);
+        Role admin = new Role();
+        admin.setRole("ADMIN");
+        roleService.add(admin);
```
let's create one admin user here too

Sofasmile commented on src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java:
```diff
@@ -44,7 +44,8 @@ public User add(User user) {
         try (Session session = sessionFactory.openSession()) {
             Query<User> getUser = session
-                    .createQuery("SELECT u FROM User u WHERE u.login = :login", User.class);
+                    .createQuery("SELECT u FROM User u left join fetch u.roles "
```
why do we use left join here? is it possible to get user without role?

Sofasmile commented on src/main/java/com/dev/cinema/model/User.java:
```diff
@@ -17,6 +19,8 @@
     private String login;
     private String password;
+    @ManyToMany
+    private List<Role> roles;
```
let's use set instead of the list

Sofasmile commented on src/main/java/com/dev/cinema/model/User.java:
```diff
@@ -42,32 +46,33 @@ public void setPassword(String password) {
     public boolean equals(Object o) {
         if (this == o) {
-            return true;
+            {
+                return true;
+            }
```
???

Sofasmile commented on src/main/java/com/dev/cinema/model/dto/UserRegistrationDto.java:
```diff
@@ -15,6 +16,8 @@
     @LoginConstraint(message = "Login is not correct")
     private String login;
+    @NotNull
+    private List<String> roles;
```
let's set USER role by default to all users at the method register()

Sofasmile commented on src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java:
```diff
@@ -1,28 +1,33 @@
-        userService.add(user);
-        shoppingCartService.registerNewShoppingCart(user);
+    public User register(User user) {
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
```
we should encode password at the userService layer
and here we can set USER role by default to all users

Sofasmile commented on src/main/java/com/dev/cinema/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public UserDetails loadUserByUsername(String login) throws UsernameNotFoundException {
+        User user = userService.findByLogin(login).orElseThrow(()
+                -> new RuntimeException("Could not find user by login" + login));
```
let's throw here "...specific Spring exception when user is not found by username"
https://mate-academy.github.io/jv-program-fulltime/04_spring/checklist/hw_08.html

Sofasmile commented on src/main/java/com/dev/cinema/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,29 @@
+        UserBuilder userBuilder = null;
+        userBuilder = org.springframework.security.core.userdetails.User.withUsername(login);
+        userBuilder.password(user.getPassword());
+        userBuilder.roles(user.getRoles().toString());
```
does `user.getRoles().toString()` work correctly?

engineer027 committed:
- fix method register() for AuthenticationServiceImpl;
- fix method loadUserByUsername() for CustomUserDetailsService;
- rename class DataInitializer and update method addRole();
- update class SecurityConfig;
- update class User;
- fix method findByLogin() for UserDaoImpl;
- fix method mapUserRegistrationDtoToUser();
- update UserRegistrationDto class;
- fix method add() for UserServiceImpl;

Sofasmile commented on src/main/java/com/dev/cinema/model/dto/UserRegistrationDto.java:
```diff
@@ -17,7 +17,7 @@
     private String login;
     @NotNull
-    private List<String> roles;
+    private Set<String> roles;
```
не варто в UserRegistrationDto приймати ролі, роль USER варто засечувати в методі /register

engineer027 committed:
update UserRegistrationDto
